src/term/mod.rs

```diff
@@ -200,6 +200,65 @@ pub use self::mode::TermMode;
 
 pub const TAB_SPACES: usize = 8;
 
+#[derive(Clone, Copy)]
+enum Charset {
+    Ascii,
+    /// A sparse charset only has a few characters in it and is best mapped with
+    /// a match statement
+    Sparse(StandardCharset),
+    /// A dense charset has most of the 96 possible characters in it and may be
+    /// better handler by a lookup table
+    Dense(StandardCharset),
+    /// A user defined charset handled by some more complicated mechanism
+    User,
+}
+
+impl Charset {
+    /// Switch/Map character to the active charset. Ascii is the common case and
+    /// for that we want to do as little as possible.
+    #[inline]
+    fn switch(&self, c: char) -> char {
+        match *self {
+            Charset::Ascii => c,
+            Charset::Sparse(set) => match set {
+                StandardCharset::LineDrawing =>
+                    match c {
```

Comment:
You can't safely use the first byte of `c`. Easiest thing to do is cast to `usize` and check that it's in the 7-bit ascii range we care about. Casting to a `usize` sets us up to index into an array (with an offset as you suggested). Although.. since we need to do some conditional logic anyway, maybe it makes sense for this character set to just use a `match`.

I would kind of like to test a couple of implementations since this will be such a hot code path. Unfortunately we're not set up to benchmark it at the moment.

Maybe for starters, just do whatever is easiest.